docs: getting started (#514)
libp2p/js-libp2p: doc/GETTING_STARTED.md

# Getting Started

Welcome to libp2p! This guide will walk you through setting up a fully functional libp2p node 🚀

- [Getting Started](#getting-started)
  - [Install](#install)
  - [Configuring libp2p](#configuring-libp2p)
    - [Basic setup](#basic-setup)
      - [Transports](#transports)
      - [Connection Encryption](#connection-encryption)
      - [Multiplexing](#multiplexing)
      - [Running Libp2p](#running-libp2p)
    - [Custom setup](#custom-setup)
      - [Peer Discovery](#peer-discovery)
      - [Pubsub](#pubsub)
  - [What is next](#what-is-next)

## Install

The first step is to install libp2p in your project:

```sh
npm install libp2p
```

## Configuring libp2p

If you're new to libp2p, we recommend configuring your node in stages, as this can make troubleshooting configuration issues much easier. In this guide, we'll do just that. If you're more experienced with libp2p, you may wish to jump to the [Configuration readme](./CONFIGURATION.md).

### Basic setup

> Now that we have libp2p installed, let's configure the minimum needed to get your node running. The only modules libp2p requires are a [**Transport**][transport] and [**Crypto**][crypto] module. However, we recommend that a basic setup should also have a [**Stream Multiplexer**](streamMuxer) configured, which we will explain shortly. Let's start by setting up a Transport.

Reviewer: Is the muxer really optional?

Reviewer 2: Yes, you should be able to run an unmuxed connection. The primary reason would be to support applications that don't need/want it, such as lower energy/power devices, although this may not be a valid use case for js-libp2p.

#### Transports

> Libp2p uses Transports to establish connections between peers over the network. You can configure any number of Transports, but you only need 1 to start with. Supporting more Transports will improve the ability of your node to speak to a larger number of nodes on the network.

Reviewer: Could we explain a little what a transport is?

You should select Transports according to the runtime of your application; Node.js or the browser. You can see a list of some of the available Transports in the [configuration readme](./CONFIGURATION.md#transport). For this guide let's install `libp2p-websockets`, as it can be used in both Node.js and the browser.

Start by installing `libp2p-websockets`:

```sh
npm install libp2p-websockets
```

Now that we have the module installed, let's configure libp2p to use the Transport. We'll use the [`Libp2p.create`](./API.md#create) method, which takes a single configuration object as its only parameter. We can add the Transport by passing it into the `modules.transport` array:

```js
const Libp2p = require('libp2p')
const WebSockets = require('libp2p-websockets')

const node = await Libp2p.create({
  modules: {
    transport: [WebSockets]
  }
})
```

> As new Transports are created, you can reevaluate the needs of your appliction and select the Transports that best suit your requirements. You may wish to remove the Transports you had before, or simply append the new Transports to `modules.transport` in order to establish connections with peers that support either.

Reviewer: ```suggestion
There are multiple libp2p transports available, you should evaluate the needs of your application and select the Transport(s) that best suit your requirements. You can add as many transports as you like to `modules.transport` in order to establish connections with as many peers as possible.
```

<details><summary>Read More</summary>
If you want to know more about libp2p transports, you should read the following content:

- https://docs.libp2p.io/concepts/transport
- https://github.com/libp2p/specs/tree/master/connections
</details>

#### Connection Encryption

Encryption is an important part of communicating on the libp2p network. Every connection must be encrypted to help ensure security for everyone. As such, Connection Encryption (Crypto) is a required component of libp2p.

> There are a growing number of Crypto modules being developed for libp2p. As those are released they will be tracked in the [available Connection Encryption](./CONFIGURATION.md#connection-encryption) section of the configuration readme. For now, we are going to configure our node to use the `libp2p-secio` module, which is widely supported across the various libp2p implementations.

Reviewer: ```suggestion
There are a growing number of Crypto modules being developed for libp2p. As those are released they will be tracked in the [Connection Encryption section of the configuration readme](./CONFIGURATION.md#connection-encryption). For now, we are going to configure our node to use the `libp2p-secio` module, which is widely supported across the various libp2p implementations.
```

- If it is likely that you will have other peers on your local network, `libp2p-mdns` is a must if you're node is not running in the browser. It allows peers to discover each other when on the same local network.
- If your application is browser based you can use the `libp2p-webrtc-star` Transport, which includes a rendezvous based peer sharing service.
- A random walk approach can be used via `libp2p-kad-dht`, to crawl the network and find new peers along the way.

For this guide we will configure `libp2p-bootstrap` as this is useful for joining the public network.

Let's install `libp2p-bootstrap`.

```sh
npm install libp2p-bootstrap
```

We can provide specific configurations for each protocol within a `config.peerDiscovery` property in the options as shown below.

```js
const Libp2p = require('libp2p')
const WebSockets = require('libp2p-websockets')
const SECIO = require('libp2p-secio')
const MPLEX = require('libp2p-mplex')

const Bootstrap = require('libp2p-bootstrap')

// Known peers addresses
const bootstrapMultiaddrs = [

>   '/dns4/ams-1.bootstrap.libp2p.io/tcp/443/wss/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd',

Reviewer: ```suggestion
  '/dns4/ams-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd',
```

>   '/dns4/lon-1.bootstrap.libp2p.io/tcp/443/wss/ipfs/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3'

Reviewer: ```suggestion
  '/dns4/lon-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3'
```

There are a growing number of Crypto modules being developed for libp2p. As those are released they will be tracked in the [available Connection Encryption](./CONFIGURATION.md#connection-encryption) section of the configuration readme. For now, we are going to configure our node to use the `libp2p-secio` module.

```sh
npm install libp2p-secio
```

With `libp2p-secio` installed, we can add it to our existing configuration by importing it and adding it to the `modules.connEncryption` array:

```js
const Libp2p = require('libp2p')
const WebSockets = require('libp2p-websockets')
const SECIO = require('libp2p-secio')

const node = await Libp2p.create({
  modules: {
    transport: [WebSockets],
    connEncryption: [SECIO]
  }
})
```

<details><summary>Read More</summary>
If you want to know more about libp2p connection encryption, you should read the following content:

- https://docs.libp2p.io/concepts/secure-comms
- https://github.com/libp2p/specs/tree/master/connections
</details>

#### Multiplexing

> With libp2p you are able to create your own protocols so that peers can communicate in a meaningful way to your application. Each protocol works as an independent logical stream and you can use a stream multiplexer, in order to share a common underlying transport medium between streams (single connection). As an example, you can create a chat protocol on top of libp2p and use it to exchange chat messages.

Reviewer 2: ```suggestion
While multiplexers are not strictly required, they are highly recommended as they improve the effectiveness and efficiency of connections for the various protocols libp2p runs. Including a multiplexer will allow libp2p to run several of its internal protocols, like Identify, as well as allowing your application to easily run any number of protocols over a single connection.
```

> Looking at the [available stream multiplexing](./CONFIGURATION.md#stream-multiplexing) protocols, we are only able to use `libp2p-mplex` for the time being. Bear in mind that future libp2p transports might have `multiplexing` capabilities already built-in (such as `QUIC`).

Reviewer 2: ```suggestion
Looking at the [available stream multiplexing](./CONFIGURATION.md#stream-multiplexing) modules, js-libp2p currently only supports `libp2p-mplex`, so we will use that here. Bear in mind that future libp2p transports might have `multiplexing` capabilities already built-in (such as `QUIC`).
```

You can install `libp2p-mplex` and add it to your libp2p node as follows in the next example.

```sh
npm install libp2p-mplex
```

```js
const Libp2p = require('libp2p')
const WEBSOCKETS = require('libp2p-websockets')
const SECIO = require('libp2p-secio')
const MPLEX = require('libp2p-mplex')

const node = await Libp2p.create({
  modules: {
    transport: [WEBSOCKETS],
    connEncryption: [SECIO],
    streamMuxer: [MPLEX]
  }
})
```

> Protocols design is similar to an http api, where each protocol has an handler, receives a message and returns a response. If you want to create your own protocol, start by having a look at the [Protocol and Stream Muxing Example](../examples/protocol-and-stream-muxing).

Reviewer 2: ```suggestion
Protocol design is similar to an http api, where each protocol has a registered handler that receives and responds to requests. If you want to create your own protocol, start by having a look at the [Protocol and Stream Muxing Example](../examples/protocol-and-stream-muxing).
```

<details><summary>Read More</summary>
If you want to know more about libp2p stream multiplexing, you should read the following content:

- https://docs.libp2p.io/concepts/stream-multiplexing
- https://github.com/libp2p/specs/tree/master/connections
- https://github.com/libp2p/specs/tree/master/mplex
</details>

#### Running Libp2p

Now that you have configured a [`Transport`][transport], [`Crypto`][crypto] and [**Stream Multiplexer**](streamMuxer) module, you can start your libp2p node. We can start and stop libp2p using the [`libp2p.start()`](./API.md#start) and [`libp2p.stop()`](./API.md#stop) methods.

```js
const Libp2p = require('libp2p')
const WebSockets = require('libp2p-websockets')
const SECIO = require('libp2p-secio')
const MPLEX = require('libp2p-mplex')

const node = await Libp2p.create({
  modules: {
    transport: [WebSockets],
    connEncryption: [SECIO],
    streamMuxer: [MPLEX]
  }
})

// start libp2p
await node.start()

// stop libp2p
await node.stop()
```

> If you are not familiar with multiaddrs, you should read the [multiformats/js-multiaddr](https://github.com/multiformats/js-multiaddr) README.

Reviewer 2: This seems out of place, not sure where this should go.

Author: I think we can just remove it

### Custom setup

> After having your libp2p node running, it is time to configure all its features according to your project needs.

Reviewer 2: ```suggestion
Once your libp2p node is running, it is time to get it connected to the public network. We can do this via peer discovery.
```

#### Peer Discovery

> In a P2P context, peer discovery is critical to a functioning system. It is not pratical to connect to all the peers we know manually, nor to get to know most of the peers' addresses in advance.

Reviewer 2: ```suggestion
Peer discovery is an important part of creating a well connected libp2p node. A static list of peers will often be used to join the network, but it's useful to couple other discovery mechanisms to ensure you're able to discover other peers that are important to your application.
```

Looking at the [available peer discovery](./CONFIGURATION.md#peer-discovery) protocols, there are several options to be considered:
- If you already know the addresses of some other network peers, you should consider using `libp2p-bootstrap` as this is the easiest way of getting your peer into the network.

> - If it is likely that you will have other peers on your network, `libp2p-mdns` is a must. It allows peers to discover each other when on the same local network with zero configuration. mDNS uses a multicast system of DNS records.

Reviewer 2: ```suggestion
- If it is likely that you will have other peers on your local network, `libp2p-mdns` is a must if you're not using the browser. It allows peers to discover each other when on the same local network with zero configuration. mDNS uses a multicast system of DNS records.
```

> - In the event of working on a browser runtime, you must go with `js-libp2p-webrtc-star`.

Reviewer 2: ```suggestion
- If your application is browser based you can use the transport `js-libp2p-webrtc-star`, which includes a rendezvous based peer sharing mechanism.
```

> - if you need to crawl the network and discover a large number of peers, `js-libp2p-kad-dht`.

Reviewer 2: ```suggestion
- A random walk approach can be used via `js-libp2p-kad-dht`, to crawl the network and find new peers along the way.
```

> Reviewing the events specification on [Libp2p.create](./API.md#create), each time a peer is discovered a `peer:discovery` event will be emitted by the node.

Reviewer 2: Should we update the API doc so that we can deep link to the Events section? I think it convolutes this section and would make it easier of we could just link directly to that events declaration there.

Author: Should we add it in the same level as the Instance Methods and Types?

Reviewer 2: Yeah, I think so

We will consider in this guide that we already know some peers and also that we are creating an application that might be used by people within the same office network. Accordingly, we will install `libp2p-bootstrap` and `libp2p-mdns`

```sh
npm install libp2p-bootstrap libp2p-mdns
```

We can provide specific configurations for each protocol within a `config.peerDiscovery` property in the options.

```js
const Libp2p = require('libp2p')

> const TCP = require('libp2p-tcp')

Reviewer 2: This switches over to TCP, any reason for that? We should keep it consistent with the rest of the guide.

  modules: {
    transport: [TCP],
    connEncryption: [SECIO],
    streamMuxer: [MPLEX],
    peerDiscovery: [Bootstrap, MulticastDNS]
  },
  config: {
    peerDiscovery: {
      autoDial: true, // Auto connect to discovered peers (limited by ConnectionManager minPeers)
      // The `tag` property will be searched when creating the instance of your Peer Discovery service.
      // The associated object, will be passed to the service when it is instantiated.
      [MulticastDNS.tag]: {
        interval: 1000,
        enabled: true
      },
      [Bootstrap.tag]: {
        interval: 20000,
        enabled: true,
        list: bootstrapers // provide array of peers
      }
    }
  }
})

node.on('peer:discovery', (peer) => {
  console.log(peer) // Log discovered peer
})

// start libp2p
await node.start()
```

<details><summary>Read More</summary>
If you want to know more about libp2p peer discovery, you should read the following content:

- https://github.com/libp2p/specs/blob/master/discovery/mdns.md
</details>

#### Peer Routing

Peer Routing offers a way to find other peers in the network by issuing queries using a Peer Routing algorithm, through their peer ids.

> Looking at the [available peer routing](./CONFIGURATION.md#peer-routing) protocols, we might choose `libp2p-kad-dht` or `libp2p-delegated-peer-routing`. If your runtime is a browser or a low power device, you might prefer to use `libp2p-delegated-peer-routing` as you are able to delegate other peer to issue the queries for you. However, you need to know a peer able to do that. Otherwise, you should use `libp2p-kad-dht`.

Reviewer 2: ```suggestion
Looking at the [available peer routing](./CONFIGURATION.md#peer-routing) modules, we might choose `libp2p-kad-dht` or `libp2p-delegated-peer-routing`. If your runtime is a browser or a low power device, you might prefer to use `libp2p-delegated-peer-routing` as you are able to delegate to another peer to issue the queries for you. However, you need to know a peer able to do that. Otherwise, you should use `libp2p-kad-dht`.
```

> In this context, we will install the dht. You should revisit the [API](./API.md) document, more specifically [API#peerRouting.findPeer](./API.md#peerRoutingfindPeer).

Reviewer 2: ```suggestion
In this guide, we will install `libp2p-kad-dht`. You should revisit the [API](./API.md) document, more specifically [API#peerRouting.findPeer](./API.md#peerRoutingfindPeer).
```

```sh
npm install libp2p-kad-dht
```

> `js-libp2p` configuration is not well defined yet regarding peer routing. Therefore, the `dht` should be plugged through a `modules.dht` property, while the `libp2p-delegated-peer-routing` should be plugged into `modules.peerRouting` array.

Reviewer 2: It is defined, but the DHT isn't currently separable into those components so it has it's own configuration parameter.

```js
const Libp2p = require('libp2p')
const TCP = require('libp2p-tcp')
const MPLEX = require('libp2p-mplex')
const SECIO = require('libp2p-secio')
const DHT = require('libp2p-kad-dht')

const node = await Libp2p.create({
  modules: {
    transport: [TCP],
    streamMuxer: [MPLEX],
    connEncryption: [SECIO],
    dht: DHT
  },
  config: {
    dht: { // The DHT options (and defaults) can be found in its documentation
      enabled: true,
      randomWalk: {
        enabled: false, // dht discovery (enabled by default)
        interval: 300e3,
        timeout: 10e3
      }
    }
  }
})

// start libp2p
await node.start()

// Find peer
const peerInfo = await libp2p.peerRouting.findPeer(peerId)

// ...
```

You can find out more about peer routing on the example provided on [./CONFIGURATION.md#setup-with-content-and-peer-routing](./CONFIGURATION.md#setup-with-content-and-peer-routing).

<details><summary>Read More</summary>
If you want to know more about libp2p peer routing, you should read the following content:

- https://docs.libp2p.io/concepts/peer-routing
</details>

#### Content Routing

> Content routing provides a way to find where content lives in the network. It works in two steps: 1) Peers provide (announce) to the network that they are holders of specific content and 2) Peers issue queries to find where that content lives. A Content Routing mechanism could be as complex as a DHT or as simple as a registry somewhere in the network.

Reviewer 2: ```suggestion
Content routing provides a way to find and announce where content lives in the network. It works in two steps: 1) Peers announce to the network that they are a provider of a certain piece of content and 2) Peers issue queries to find where that content lives. A Content Routing mechanism could be as complex as a DHT or as simple as a registry somewhere in the network.
```

> Looking at the [available content routing](./CONFIGURATION.md#content-routing) protocols, we might choose `libp2p-kad-dht` or `libp2p-delegated-content-routing`. If of your runtime is a browser or a low power device, you might prefer to use `libp2p-delegated-peer-routing` as you are able to delegate other peer to issue the queries for you. However, you need to know a peer able to do that. Otherwise, you should use `libp2p-kad-dht`.

Reviewer 2: ```suggestion
Looking at the [available content routing](./CONFIGURATION.md#content-routing) modules, we might choose `libp2p-kad-dht` or `libp2p-delegated-content-routing`. If your runtime is a browser or a low power device, you might prefer to use `libp2p-delegated-content-routing` as you are able to delegate to another peer to issue the queries for you. However, you need to know a peer able to do that. Otherwise, you should use `libp2p-kad-dht`.
```

> In this context, we will install the dht. You should revisit the [API](./API.md) document, more specifically [API#contentrouting.findProviders](./API.md#contentroutingfindproviders), [API#contentrouting.provide](./API.md#contentroutingprovide), [API#contentrouting.put](./API.md#contentroutingput), [API#contentrouting.get](./API.md#contentroutingget) and [API#contentrouting.contentrouting.getMany](./API.md#contentroutingcontentroutinggetmany).

Reviewer 2: ```suggestion
In this guide, we will install `libp2p-kad-dht`. If you'd like to review the available content routing methods, you should revisit the [API](./API.md) document, more specifically [API#contentrouting.findProviders](./API.md#contentroutingfindproviders), [API#contentrouting.provide](./API.md#contentroutingprovide), [API#contentrouting.put](./API.md#contentroutingput), [API#contentrouting.get](./API.md#contentroutingget) and [API#contentrouting.contentrouting.getMany](./API.md#contentroutingcontentroutinggetmany).
```

```sh
npm install libp2p-kad-dht
```

> `js-libp2p` configuration is not well defined yet regarding content routing routing. Therefore, the `dht` should be plugged through a `modules.dht` property, while the `libp2p-delegated-content-routing` should be plugged into `modules.contentRouting` array.

Reviewer 2: Same here, it is defined, but the DHT does not yet conform to the separate configuration.

```js
const Libp2p = require('libp2p')
const TCP = require('libp2p-tcp')
const MPLEX = require('libp2p-mplex')
const SECIO = require('libp2p-secio')
const DHT = require('libp2p-kad-dht')

const node = await Libp2p.create({
  modules: {
    transport: [TCP],
    streamMuxer: [MPLEX],
    connEncryption: [SECIO],
    dht: DHT
  },
  config: {
    dht: { // The DHT options (and defaults) can be found in its documentation
      enabled: true,
      randomWalk: {
        enabled: false, // dht discovery (enabled by default)
        interval: 300e3,
        timeout: 10e3
      }
    }
  }
})

// start libp2p
await node.start()

// provide cid

> await libp2p.contentRouting.provide(cid)

Reviewer 2: Since we're aiming to have people join the live network we should probably not have them provide content, as their node is going to go offline pretty quickly. What do you think?

For findProviders we should just use a well known cid, and we should include the cid module here. A website or the ipfs readme might be good options.

Author: Good call!

// get providers for a cid
for await (const provider of libp2p.contentRouting.findProviders(cid2)) {
  console.log(provider)
}

// ...
```

If you are not familiar with `CIDs`, you should have a look at [multiformats/js-cid](https://github.com/multiformats/js-cid). Moreover, you can find out more about peer routing on the example provided on [./CONFIGURATION.md#setup-with-content-and-peer-routing](./CONFIGURATION.md#setup-with-content-and-peer-routing).

<details><summary>Read More</summary>
If you want to know more about libp2p content routing, you should read the following content:

- https://docs.libp2p.io/concepts/content-routing
</details>

### Pubsub

> If you are looking for real time message exchange between peers, pubsub may be what you are looking for. Publish/Subscribe is a system where peers congregate around topics they are interested in. Peers interested in a topic are said to be subscribed to that topic and should receive the data published on it from other peers.

Reviewer 2: ```suggestion
If you are looking for real time message exchange between multiple peers, pubsub may be what you are looking for. Publish/Subscribe is a system where peers congregate around topics they are interested in. Peers interested in a topic will subscribe to that topic and receive the data published on it from other peers.
```

> Looking at the [available pubsub routers](./CONFIGURATION.md#pubsub), you might choose `libp2p-gossipsub` first, and consider other options later on. You should revisit the [API](./API.md) document, more specifically [API#pubsub.getSubscribers](./API.md#pubsubgetsubscribers), [API#pubsub.getTopics](./API.md#pubsubgettopics), [API#pubsub.publish](./API.md#pubsubpublish) and [API#pubsub.subscribe](./API.md#pubsubsubscribe).

Reviewer 2: ```suggestion
You can see the [available pubsub routers](./CONFIGURATION.md#pubsub) in the configuration readme. In this guide we will use `libp2p-gossipsub` as it is a more efficient alternative to `libp2p-floodsub`. You can revisit the [API](./API.md) document to familiarize yourself with the available methods, more specifically [API#pubsub.getSubscribers](./API.md#pubsubgetsubscribers), [API#pubsub.getTopics](./API.md#pubsubgettopics), [API#pubsub.publish](./API.md#pubsubpublish) and [API#pubsub.subscribe](./API.md#pubsubsubscribe).
```

```sh
npm install libp2p-gossipsub
```

```js
const Libp2p = require('libp2p')
const TCP = require('libp2p-tcp')
const MPLEX = require('libp2p-mplex')
const SECIO = require('libp2p-secio')
const GossipSub = require('libp2p-gossipsub')

const node = await Libp2p.create({
  modules: {

>     transport: [TCP],

Reviewer 2: Same thing here, why the switch to TCP?

      - [Peer Routing](#peer-routing)
      - [Content Routing](#content-routing)
    - [Pubsub](#pubsub)
  - [What is next](#what-is-next)

## Install

The first step is to install libp2p in your project:

```sh
npm install libp2p
```

## Configuring libp2p

If you're new to libp2p, we recommend configuring your node in stages, as this can make troubleshooting configuration issues much easier. In this guide, we'll do just that. If you're more experienced with libp2p, you may wish to jump to the [Configuration readme](./CONFIGURATION.md).

### Basic setup

Now that we have libp2p installed, let's configure the absolute minimum needed to get your node running. The only modules libp2p requires are a [**Transport**][transport] and [**Crypto**][crypto] module. Moreover, we recommend that a basic setup should also have a [**Stream Multiplexer**](streamMuxer) configured. Let's start by setting up a Transport.

#### Transports

Libp2p uses Transports to establish connections between peers over the network. You can configure 1 Transport, or as many as you like. Supporting more Transports will improve the ability for other nodes on the network to communicate with you.

> You should select Transports according to the runtime where your application will run. You can see a list of some of the available Transports in the [configuration readme](./CONFIGURATION.md#transport). We are going to install `libp2p-websockets`, as it can be used in both Node.js and the browser.

Reviewer 2: ```suggestion
You should select Transports according to the runtime of your application (Node.js / browser). You can see a list of some of the available Transports in the [configuration readme](./CONFIGURATION.md#transport). We are going to install `libp2p-websockets`, as it can be used in both Node.js and the browser.
```

There are a growing number of Crypto modules being developed for libp2p. As those are released they will be tracked in the [available Connection Encryption](./CONFIGURATION.md#connection-encryption) section of the configuration readme. For now, we are going to configure our node to use the `libp2p-secio` module.

```sh
npm install libp2p-secio
```

With `libp2p-secio` installed, we can add it to our existing configuration by importing it and adding it to the `modules.connEncryption` array:

```js
const Libp2p = require('libp2p')
const WebSockets = require('libp2p-websockets')
const SECIO = require('libp2p-secio')

const node = await Libp2p.create({
  modules: {
    transport: [WebSockets],
    connEncryption: [SECIO]
  }
})
```

<details><summary>Read More</summary>
If you want to know more about libp2p connection encryption, you should read the following content:

- https://docs.libp2p.io/concepts/secure-comms
- https://github.com/libp2p/specs/tree/master/connections
</details>

> #### Multiplexing

Reviewer 2: I think we need to split multiplexing configuration and protocol creation into separate sections.

This Multiplexing section also feels like it's diving a little too deep into what multiplexing is. I think we should reduce the up front language and the leverage the `Read More` section to  send folks to more detailed information.

Reviewer 2: I think the protocol creation section should also come after starting and stopping your node. That way if people have issues with their early configuration, it can be caught prior to building out a custom protocol.

Do you think it would be more helpful to point people to the protocol example, instead of building that out here? I think `Getting Started` can focus on creating a first node, and connecting it to the public network with basic peer discovery and new connection logging. Then we could point to the examples folder. This would also put us in a good spot to revamp those in the future, where all examples are based off of the configuration from the getting started guide.

More than connecting with other peers, you want to have them communicating in a meaningful way to your application. With libp2p you are able to create your own protocols so that you can integrate your business logic into the network. As an example, you can create a chat protocol on top of libp2p and use it to exchange chat messages.

In this context, your node will not be efficient if you open multiple connections with a single peer to use different protocols. Therefore, each protocol works as an independent logical stream and you can use a stream multiplexer, in order to share a common underlying transport medium between streams (single connection).

Looking at the [available stream multiplexing](./CONFIGURATION.md#stream-multiplexing) protocols, we are only able to use `libp2p-mplex` for the time being. Bear in mind that future libp2p transports might have `multiplexing` capabilities already built-in (such as `QUIC`).

You can install `libp2p-mplex` and add it to your libp2p node as follows in the next example. You should revisit the [API](./API.md) document, more specifically [API#dial](./API.md#dial) and [API#handle](./API.md#handle).

> We will also install the next dependencies for easily work with [async iterable duplex streams](https://gist.github.com/alanshaw/591dc7dd54e4f99338a347ef568d6ee9).

Reviewer 2: We should point to the local README we have for this, as we'll also be including more examples of using iterable streams in libp2p https://github.com/libp2p/js-libp2p/blob/refactor/async-await/doc/STREAMING_ITERABLES.md
